junipervpn/vpn.py

```diff
@@ -65,6 +68,110 @@ def hotp(key):
     counter = int2beint64(int(time.time()) / 30)
     return dec(hmac.new(key, counter, hashlib.sha256).digest(), 6)
 
+
+class NetworkMonitorThread(threading.Thread):
+    """
+    :param rate: Number of packets emitted per second
+    :type rate: float
+
+    :param on_disconnect: Callable to call when a network disconnection is
+        detected.
+    :type on_disconnect: collections.abc.Callable
+
+    :param host: Host to ping.
+    :type host: str
+
+    :param ping: Ping command to use.
+    :type ping: str
+
+    :param ping_rate: Ping rate in packet/s
+    :type ping_rate: float
+
+    :param ping_duration: Duration of each ping command. A short-enough interval is
+        preferred, so that the script can react quickly to network outage
+    :type ping_duration: int
+
+    :param connect_timeout: Time to wait in seconds between reconnection
+        attempts.
+    :type connect_timeout: float
+
+    :param daemon: Passed to threading.Thread.
+    :type daemon: bool
+    """
+    def __init__(self,
+        on_disconnect,
+        host='google.com',
+        ping='ping',
+        connect_timeout=10,
+        ping_rate=1,
+        ping_duration=2,
+        daemon=True,
+    ):
+        self.host = host
+        self.ping = ping
+        self.ping_rate = ping_rate
+        self.ping_duration = ping_duration
+        self.connect_timeout = connect_timeout
+        self.on_disconnect = on_disconnect
+
+        self._stop = threading.Event()
+        super().__init__(name='ping_network_monitor', daemon=daemon)
+
+    def run(self):
+        interval = 1 / self.ping_rate
+        # interval < 0.2 is not allowed as regular user, so best avoided
+        interval = interval if interval >= 0.2 else 0.2
+
+        deadline = int(math.ceil(self.ping_duration))
+        cmd = ['ping', '-q', '-w', str(deadline), '-i', str(interval), '--', self.host]
+
+        # Exit the thread if it needs to be stopped.
+        # It is executed before attempting any long command
+        def check_stop():
+            if self._stop.is_set():
+                sys.exit(0)
```

Comment:
I know `sys.exit` is correct, but I had to google it ;) Does using `thread.exit()` work instead, to make it a bit clearer?

Reply:
`thread.exit()` does not exist in Python 3. The module has been renamed to `threading` and does not include any `exit()` function. The only exit-related documented bit is:
```
If exc_type is SystemExit, the exception is silently ignored. Otherwise, the exception is printed out on sys.stderr.
```
https://docs.python.org/3/library/threading.html